utils/src/main/kotlin/LicenseFilenamePatterns.kt

```diff
@@ -44,4 +46,65 @@ object LicenseFilenamePatterns {
     val ROOT_LICENSE_FILENAMES = listOf(
         "readme*"
     ).generateCapitalizationVariants()
+
+    /**
+     * Return glob patterns which match all files which may contain license information residing recursively within the
+     * given absolute [directory] or in any of its ancestor directories.
+     */
+    fun getLicenseFileGlobsForDirectory(directory: String): List<String> =
+        globFileInDirectoryOrAncestors(directory, LICENSE_FILENAMES + ROOT_LICENSE_FILENAMES)
+
+    /**
+     * Return recursively all ancestor directories of the given absolute [directory].
+     */
+    private fun getAllAncestorDirectories(directory: String): List<String> {
+        val result = mutableListOf<String>()
+
+        var ancestorDir = File(directory).parentFile
+        while (ancestorDir != null) {
+            result.add(ancestorDir.invariantSeparatorsPath)
+            ancestorDir = ancestorDir.parentFile
+        }
+
+        return result
+    }
+
+    /**
+     * Return a glob pattern which matches files in [directory], in any ancestor directory of [directory] and
+     * optionally (if [matchSubDirs] is true) recursively in any sub-directory of directory, if the filename
+     * matches the [filenamePattern].
+     */
+    internal fun globFileInDirectory(
+        directory: String,
+        filenamePattern: String,
+        matchSubDirs: Boolean
+    ): String {
+        val separator = "/".takeIf { !directory.endsWith("/") }.orEmpty()
+        val globStar = "**/".takeIf { matchSubDirs }.orEmpty()
```

Comment:
Mabye `matchSubDirs && !filenamePattern.startsWith("**/")`?

Reply:
I prefer not to do that as file name patterns are currently just for matching file names, but not for matching directories. An I deliberately want to keep that - because changing it would prevent re-use. That re-usability I want to keep and actually do need in one of my following PRs.